### ISSUE TITLE
Error deploying with cloudsql connection

### ISSUE THREAD

**Issue by aronsogor** [GoogleCloudPlatform/cloud-code-vscode#613]
**Environment**:
*(Hint: "Report Extension Issue on Github" command will fill these out for you.)*

Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)

**Version information**

Cloud Code Extension version:
Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)
VSCode version:
Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)
OS: 
Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)
Cloud SDK:
Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)
Skaffold: 
Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)
Kubectl:
Cloud Shell Editor v1.25.0
(Built with Eclipse Theia)
**Description**:
Error deploying with cloudsql connection

Running gcloud command: /usr/bin/gcloud beta code export --readiness-probe --service-name anyrolegcp --image anyrolegcp --no-skaffold-file --dockerfile= --kubernetes-file /tmp/cloud-code-cloud-run-MKJCsq/pods_and_services.yaml --env-vars DB_USER=dev-db,INSTANCE_CONNECTION_NAME=anyroleon:us-west1:anyrole-one,DB_NAME=anyrole,K_SERVICE=anyrolegcp,K_REVISION=local,K_CONFIGURATION=anyrolegcp --cloudsql-instances anyroleon:us-west1:anyrole-one --service-account dev-db@anyroleon.iam.gserviceaccount.com --memory 256Mi --quiet
ERROR: (gcloud.beta.code.export) unrecognized arguments: --cloudsql-instances 

**Repro step**:
Cloud Code: Run on Cloud Run Emulator 
 
specify a CloudSQL instance and the build will fail. How do I get to connect to Cloudsql Emulator?

**Comment by davidponder** [comment]
Thank you for bringing this to our attention! We're going to investigate this and get back to you ASAP!

Thanks again!

**Comment by aronsogor** [comment]
You're welcome! Happy to give you any diagnostic, it pretty consistently
reproduceable

**Comment by aronsogor** [comment]
I am wondering, can someone explain what will the fix do?

Specifically, I am guessing that the cloud SQL proxy configuration will get generated and it will be bootstrapped into the emulator docker image. Is there a way to log in to the emulator image and start the cloud SQL manually? Is there a template file I can just replace to config and rig it to some startup script in the instance (which is what you probably do)

Or can you just point me to the source code where I could learn how this command was intended or used to work?

The current documentation is pretty hard to follow to start with. I would like to get a version of the deployment work where:
- Cloud Run runs under a service account - "myrun"
- "myron" is a provisioned user on the MySQL instance
- The cloudrun code can grab the "myrun" auth token from runtime in the emulator/deployment and use it as DB_PASSWD to login to the mysql instance.

I think it should work and this way there is no need to create a password, put it into the secret manager, and deployment is a lot more elegant. One less password laying around to be missmanaged.

I am happy to help, would you work with me?

Aron

**Comment by SKrupa** [comment]
Currently, the extension is using the beta gcloud functionality, which does not yet support cloudsql instances. The gcloud alpha does allow for cloudsql, so we will be updating our extension to call the alpha API in these cases.

Once we switch over to the new API, the flow you outlined should be correct; however, this authentication flow requires a couple more changes in the extension code to fully enable. I'll update this thread once we are ready to release this.

As for a workaround, I'm taking a look to see if creating a config file manually is an option and will update you with details once I've learned more.

**Comment by SKrupa** [comment]
Hi Aron, I've looked into this a bit more, and for the local Run/Debug flow, you won't necessarily need to deal with secret manager; Cloud Code will create a certificate on the deployment target that will be used for connecting to the DB.

For example, based on this example:
https://github.com/GoogleCloudPlatform/cloud-sql-python-connector

one is able to connect via
conn: pymysql.connections.Connection = connector.connect(
            <project:region:database>,
            "pymysql",
            user="root",
            db=<your-db-name>
        )

without providing a password, and the certificate will be used.

We're still working on getting the changes in the extension to enable this flow.

**Comment by SKrupa** [comment]
We will be releasing a fix for this in the next release. With this initial fix, you will need to set the database credentials as an environment variable in the UI or in your launch.json. We will be enabling Secret Manager in a later release to allow referencing of secrets instead of having to provide a plaintext password.

**Comment by aronsogor** [comment]
SKrupa,

I am a little puzzled. Why not just use a service account?

- If I am running the app under a service account I can grab an auth token
- If that service account exists in MySql than the password is the auth token

In this case no need to have any passwords stored anywhere. Am I missing something?
I feel confused

**Comment by SKrupa** [comment]
Sorry for the confusion, I had gotten a bit mixed up with what is supported for Cloud SQL in general and what is currently supported for this use case which is connecting to the instance from within the Cloud Run local run emulator. The current integration between Cloud Run local runs and Cloud SQL is in alpha state, and some parts of the integration are still being developed and have some limitations.

Cloud SQL does allow for grabbing of the auth token in the manner you described so that passing credentials manually is not required; however, the way Cloud Run currently connects to the Cloud SQL database does not allow for this without more changes in the local cloud run emulator.

I'm following up with the team to make sure this limitation is being addressed so that ultimately users will be able to just specify a properly provisioned service account and a Cloud SQL connection and not have to set up any additional secrets. 

In the meantime, the only options are to either pass the password as an environment var as part of the launch config, or to use a database without a password set on the root account. I've included this github issue in our internal tracking items and we have a better idea of when the authentication experience can be improved. Thank you for your patience!
